Fixed FP to FP8 downcast

AndreyPavlenko opened the pull request:
Fixes #4630

This implementation is based on an arithmetic approach (see below) and provides an RTNE conversion from F32/16 to F8 types. Despite it is using more expensive fp binary operations, I don't see any performance degradation (you can find a simple bench commented out in the test). The implementation is very simple and template-based. It replaces 4 existing functions that are more complicated and adds a direct conversion from F32 to F8 without intermediate F16.

The implementation is based on the following approach. To convert a float from src to dst, we need to find such an exponent and mantissa that src is equal or close to dst. For the normal numbers (exp != 0) it means:
```python
2^(srcExp - SrcBias) * (1 + srcMan/2^SrcMBits) = 2^(dstExp - DstBias) * (1 + dstMan/2^DstMBits)
```
and the following for subnormals (dstExp == 0):
```python
src = 2^(1 - DstBias) * (dstMan/2^DstMBits)
```
The exponent is calculated as:
```python
dstExp = max(0, srcExp - SrcBias + DstBias)
```
Simplifying the first formula, we can get the following:
```python
dstMan = srcMan * 2^(DstMBits - SrcMBits)
```
If `SrcBias == DstBias` (in the case of FP16 to F8E5 conversion), this formula also works for subnormals. In the general case, we can use the following formula for subnormals:
```python
dstMan = src * 2^(DstMBits + DstBias - 1)
```
Thus, to get the mantissa, we are using a simple multiplication by a constant and rounding the result to the nearest int. In case of the mantissa overflow, we need to reset it to zero and increment the exponent. It could be done in the following way:
```python
dst = max((dstExp << DstMBits) + dstMan, DST_MAX)
```
It gives us the required dst value without the sign bit.

AndreyPavlenko commented on python/test/unit/intel/test_conversions.py:
```diff
@@ -0,0 +1,69 @@
+    dst = torch.empty_like(src, dtype=dst_dtype, device=device)
+
+    def run_kernel():
+        type_convert[(SIZE // BLOCK_SIZE, )](triton.reinterpret(src, src_dtype), triton.reinterpret(dst, dst_dtype),
```
This test is based on the upstream https://github.com/intel/intel-xpu-backend-for-triton/blob/main/python/test/unit/language/test_conversions.py#L455 . The parameter names and the min/max definitions are copied from there.

AndreyPavlenko replied:
Without it, the test fails with `TypeError: Unsupported type: <class 'torch.dtype'>`

LiyangLingIntel commented:
> > I think this at least needs to go through the benchmark set that  have.
> 
> Here are the bench results - https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16470155352/job/46557072547

It's not obvious to see the performance change from the existing benchmark. Can you add the fp to fp8 cases to [benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py](https://github.com/intel/intel-xpu-backend-for-triton/blob/main/benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py)? So that we can compare the downcasting performance before and after this change dirrectly.

AndreyPavlenko commented on third_party/intel/lib/TritonIntelGPUToLLVM/ElementwiseOpToLLVM.cpp:
```diff
@@ -524,60 +442,69 @@ static SmallVector<Value> Fp16_to_Fp8E4M3Nv(Location loc,
+      std::is_same_v<DstTy, Float8E5M2Type> ? 0x7B : DST_NAN - 1;
+  constexpr bool IS_FP16 = std::is_same_v<SrcTy, Float16Type>;
+  TritonLLVMIRRewriter b(loc, rewriter);
+  auto srcTy = IS_FP16 ? f16_ty : f32_ty;
```
It's mentioned below, that bf16 is converted to fp32, because `llvm.nearbyint` does not support bf16.

AndreyPavlenko commented:
> Can you add the fp to fp8 cases to [benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py](https://github.com/intel/intel-xpu-backend-for-triton/blob/main/benchmarks/micro_benchmarks/conversion/float_conversion/float_conversion.py)?

Done. I've also modified the benchmark and replaced the backward conversions to fp32 in the loop with bitcasts and arithmetics. Performance impact from these operations should be comparable for all the benchmarks, unlike the conversion to fp32, which is different for each type and, for example, in case of fp8e4->fp32 the impact is significant.

Here are the results:
Before (https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16583834920/job/46905289067):
```TSV
            N  fp16->fp8e4nv/rtne  fp16->fp8e5/rtne  bf16->fp8e4nv/rtne  bf16->fp8e5/rtne  fp32->fp8e4nv/rtne  fp32->fp8e4nv/rtz  fp32->fp8e5/rtne  fp32->fp8e5/rtz  fp32->fp16/rtne  fp32->fp16/rtz  fp32->bf16/rtne  fp32->bf16/rtz
0      4096.0            0.024335          0.071409            0.020754          0.021179            0.028802           0.200196          0.052232         0.347119         0.431158        0.409600         0.435745        0.136170
1     16384.0            0.096105          0.282094            0.082199          0.084124            0.113447           0.799220          0.205468         1.388475         1.710230        1.644980         1.721008        0.541441
2     65536.0            0.384060          1.124502            0.328402          0.336220            0.453285           3.187549          0.820431         5.461333         6.812474        6.514513         6.812474        2.161478
3    262144.0            1.533723          4.488767            1.312032          1.344328            1.811888          12.663961          3.279259        21.772757        26.914168       25.852465        27.025155        8.611827
4   1048576.0            5.807355         11.358059            4.205406          4.244560            7.097441          37.130878         12.400378        69.626560        92.630389       87.381333        91.180522       30.183535
5   4194304.0           11.466740         17.374913            6.930443          6.981663           21.496023          61.590367         31.013783       122.928019       155.459748      153.188605       157.918072       52.415696
6  16777216.0           11.705633         17.926291            7.017700          7.070939           22.303010          66.057233         32.718156       142.639143       205.477232      184.243532       208.983757       57.258169
7  67108864.0           11.965820         18.072081            7.039962          7.094290           22.612251          67.263570         33.203800       156.430918       231.905674      206.851598       218.240208       59.060147
```

After (https://github.com/intel/intel-xpu-backend-for-triton/actions/runs/16583855278/job/46905344649):
```TSV
            N  fp16->fp8e4nv/rtne  fp16->fp8e5/rtne  bf16->fp8e4nv/rtne  bf16->fp8e5/rtne  fp32->fp8e4nv/rtne  fp32->fp8e4nv/rtz  fp32->fp8e5/rtne  fp32->fp8e5/rtz  fp32->fp16/rtne  fp32->fp16/rtz  fp32->bf16/rtne  fp32->bf16/rtz
0      4096.0            0.029750          0.040474            0.031517          0.031527            0.046572           0.200784          0.046577         0.347119         0.431158        0.411245         0.432068        0.135809
1     16384.0            0.118725          0.161642            0.124461          0.124461            0.186013           0.801566          0.186055         1.388475         1.713808        1.648290         1.717400        0.542158
2     65536.0            0.474623          0.646057            0.497389          0.497540            0.743629           3.196878          0.743713         5.516498         6.812474        6.553600         6.826667        2.164333
3    262144.0            1.894652          2.576101            1.986541          1.985939            2.967444          12.676209          2.968116        21.772757        26.804090       25.750884        26.694908        8.611827
4   1048576.0            7.011072          9.525581            6.795697          6.793055           10.877344          36.663497         10.877344        68.894612        91.980351       87.381333        91.498778       29.993593
5   4194304.0           13.997811         17.944314           11.900760         11.896710           25.420024          61.464009         25.438525       122.784075       155.922082      153.076788       157.562134       52.402599
6  16777216.0           14.349312         18.512994           12.149303         12.149655           26.614448          66.070240         26.599680       142.590651       206.209636      184.081808       208.697798       57.307064
7  67108864.0           14.455391         18.678501           12.219430         12.220097           26.934693          67.263570         26.937287       154.468556       213.389500      206.908997       217.222969       59.071584
```
There is a slight performance degradation in fp32->fp8e5, but in all other cases the new implementation is faster, especially in case of bf16. Despite the new impl is slower for fp32->fp8e5, it handles properly the corner cases with min/max/inf/nan and the tests #4630 pass. I'm not sure if the old impl will be still faster, when fixed.

AndreyPavlenko commented on python/test/unit/intel/test_conversions.py:
```diff
@@ -0,0 +1,63 @@
+
+    # import timeit
+    # time = timeit.timeit(run_kernel, number=100000)
+    # print(f"{src_dtype} -> {dst_dtype} time: {time}.")
```
Done.
